### ISSUE TITLE
Physics simulation changes significantly when a rigidbody is selected in Scene view

### ISSUE THREAD

**Issue by VoxelBoy** [Phong13/BulletSharpUnity3d#28]
Video of the problem: https://dl.dropboxusercontent.com/u/561110/ballThrowTest.mp4

The scene has a very simple setup. 3 static box colliders and 1 rigidbody sphere. On Start, I wait until the sphere rigidbody isInWorld, and then add one-time force to it. The resulting simulation changes depending on whether the sphere rigidbody is selected or not.

I've seen before that selecting an object in Unity can have a significant impact on Editor FPS; maybe this is the root cause. However, I've tested the same setup with PhysX and the problem doesn't occur there.

**Comment by Phong13** [comment]
Interesting. Bullet Physics is supposed to be deterministic when run on the same hardware and the same build. I will try the test to see if I can reproduce the problem.

**Comment by Phong13** [comment]
I was not handling the deltaTime parameter being passed into StepSimulation correctly. Please try it again and let me know if the problem is fixed now.

**Comment by VoxelBoy** [comment]
I haven't seen a commit from you related to this issue. Am I missing something?

**Comment by Phong13** [comment]
Opps. Pushed the commit.

**Comment by VoxelBoy** [comment]
The issue still persists.
Here's a unitypackage of the scene shown in the video I linked earlier:
https://dl.dropboxusercontent.com/u/561110/ballThrowTest.unitypackage
(Import it into a Unity project that already has the Bullet library files, it's just one scene and one script)

You can use it to see if the problem is resolved.

**Comment by Phong13** [comment]
Thanks.

On Fri, Jun 10, 2016 at 6:43 PM, Yilmaz Kiymaz <notifications@github.com>
wrote:

> The issue still persists.
> Here's a unitypackage of the scene shown in the video I linked earlier:
> https://dl.dropboxusercontent.com/u/561110/ballThrowTest.unitypackage
> (Import it into a Unity project that already has the Bullet library files,
> it's just one scene and one script)
>
> You can use it to see if the problem is resolved.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Phong13/BulletSharpUnity3d/issues/28#issuecomment-225329891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AK1VEWyyTBM-Zbrws75gzgvP1HWBVZgnks5qKhLIgaJpZM4IyuxY>
> .
>



-- 
Ian Deane
DigitalOpus <http://digitalopus.ca>
Twitter: @DigitalOpus <https://twitter.com/DigitalOpus>
Vote for my new game EVA Infinity on Steam Greenlight
<http://steamcommunity.com/sharedfiles/filedetails/?id=535754554>!

**Comment by Phong13** [comment]
I spent some time looking at the scene. If the ball is not selected then the simulation is deterministic over multiple runs which is good. If the ball is selected then the simulation is affected. If I collapse the BRigidBody inspector then the simulation runs correctly. The editor has code like:

rb.friction = EditorInterface.Layout.DrawFloat("Friction", rb.friction, rb);

Which reads the friction from the rigid body and writes it back to the rigid body. Effectively we are setting many of the fields on the rigid body (mass, linearFactor, friction, etc...). It seems that assigning these properties (even reassigning the same value) is affecting the simulation.

Looks like we need to rewrite the inspector so that we only assign a value to the rigid body if it is different.

**Comment by VoxelBoy** [comment]
Good find! In OnInspectorGUI, you can use GUI.changed or EditorGUI.BeginChangeCheck to check if user has changed any parameters, and only then write the values back to the rigidbody.

**Comment by VoxelBoy** [comment]
Upon further examination, it almost looks like values such as angularDrag aren't written to the bullet rigidbody **at all** if the rigidbody inspector isn't visible. For example, with the inspector visible, if I set the angularDrag to some high value such as 200, the ball stops pretty fast after it hits the ground. But with the object **unselected** and the angularDrag still 200, the simulation behaves as if angularDrag is 0.

**Comment by Phong13** [comment]
I fixed the BRigidBody inspector so it only writes values to the bullet object if something has changed. This fixes the first bug mentioned.  This will need to be done for other inspectors such as the BSoftBody inspector too.

I was not able to reproduce the other problem you mentioned. I checked the BRigidBody class and angularDamping is a property that that writes to the bullet object if it exists. I tested it and it worked as expected both through the inspector and through a script that assigned an angularDamping value.
